_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,673 @@
+takes an input value, classifies it, and then jumps to code written to
+handle the identified class of data.
+
+In this post we explore how Rust processes such data via `match`.
+The crucial elements that `match` and its counterpart `enum` tie
+together are:
+
+* Structural pattern matching of algebraic data types: The `match`
+  construct enables case analysis with ergonomics that are vastly
+  improved over that provided by a C or Java style `switch` statement.
+
+* Exhaustive case analysis, which ensures that no case is omitted
+  when processing an input.
+
+* `match` embraces both imperative and functional styles of
+  programming: The compiler's static analyses work hard to ensure
+  statement-oriented programming remains palatable, rather than
+  forcing everyone to adopt an expression-oriented mindset.
+
+* Destructuring bind of *L-values*: Rust encourages the developer to
+  think carefully about ownership and borrowing. To ensure that
+  processing data does not force one to give up ownership of a value
+  prematurely, `match` is designed with support for merely *borrowing*
+  substructure within its input (as opposed to always *moving* such
+  substructure).
+
+We cover each of the items above in detail below, but first we
+establish a foundation for the discussion: What does `match` look
+like, and how does it work?
+
+### The Basics of `match`
+
+The `match` expression in Rust has this form:
+
+```rust
+match INPUT_EXPRESSION {
+    PREDICATE_1 => RESULT_EXPRESSION_1,
+    PREDICATE_2 => RESULT_EXPRESSION_2,
+    ...
+    PREDICATE_n => RESULT_EXPRESSION_n
+}
+```
+
+where each of the `PREDICATE_i` contains at least one *pattern*. A
+pattern describes a subset of the possible values to which
+`INPUT_EXPRESSION` could evaluate.
+The syntax `PREDICATE => RESULT_EXPRESSION` is called a "match arm",
+or simply "arm".
+
+Patterns can match atomic values, like integers or characters; they
+can also match user-defined symbolic data, defined via `enum`.
+
+The below code demonstrates generating the next guess (poorly) in a number
+guessing game, given the answer from a previous guess.
+
+```rust
+enum Answer {
+    Higher,
+    Lower,
+    Bingo,
+}
+
+fn suggest_guess(prior_guess: u32, answer: Answer) {
+    match answer {
+        Answer::Higher => println!("maybe try {} next", prior_guess + 10),
+        Answer::Lower  => println!("maybe try {} next", prior_guess - 1),
+        Answer::Bingo  => println!("we won with {}!", prior_guess),
+    }
+}
+
+#[test]
+fn demo_suggest_guess() {
+	suggest_guess(10, Answer::Higher);
+	suggest_guess(20, Answer::Lower);
+	suggest_guess(19, Answer::Bingo);
+}
+```
+
+(Incidentally, nearly all the code in this post is directly
+executable; you can cut-and-paste the code snippets into a file
+`demo.rs`, compile the file with `--test`, and run the resulting
+binary to see the tests run.)
+
+Patterns can also match structured data (e.g. tuples, slices, user-defined
```

Comment:
a reviewer in `#rust` says this paragraph may be too terminology heavy.

Also, at this point I don't think I've introduced side-conditions, so that text itself does not actually make sense.  Which opens the door for mucho simplification.